### PR TITLE
Node 669 group extensions

### LINKED ISSUE #669: Group Extensions Migration
Prepare database schema for group extensions

### PR DESCRIPTION
**Pull request by thomasgwatson**
Closes #669 and #672 

Got some good review time in with Tibet this afternoon.

### CODE REVIEW

File: api/models/Extension.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module.exports = bookshelf.Model.extend(Object.assign({
+    tableName: 'extensions',
```
Comment (tibetsprague) [inline api/models/Extension.js:2]: indentation looks off here

File: api/models/GroupExtension.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module.exports = bookshelf.Model.extend(Object.assign({
+    tableName: 'group_extensions',
```
Comment (tibetsprague) [inline api/models/GroupExtension.js:2]: indentation looks off

File: api/models/GroupExtension.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    hasTimestamps: true,
+  
+    group () {
+      return this.belongsTo(Group)
+    },
+  
+    extension () {
+      return this.belongsTo(Extension)
+    },
+    async type () {
```
Comment (tibetsprague) [inline api/models/GroupExtension.js:13]: missing a newline, shouldnt the linter catch these things?

File: migrations/20210914160014_group_extensions.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+exports.up = async function(knex) {
+    await knex.schema.createTable('extensions', table => {
```
Comment (tibetsprague) [inline migrations/20210914160014_group_extensions.js:2]: hmm, why does the indentation seem off for all these files? something with your setup? should all be two spaces not 4

File: migrations/20210914160014_group_extensions.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        table.jsonb('data')
+        table.boolean('active').defaultsTo(true)
+        table.index('group_id')
+        table.timestamp('created_at')
+        table.timestamp('updated_at')
+    })
+
+
+    await knex.schema.alterTable('groups', async (table) => {
+      table.text('type').defaultsTo('group')
```
Comment (tibetsprague) [inline migrations/20210914160014_group_extensions.js:22]: this is a good question, do we always have a type? a default of 'group' seems kind of redundant and unecessary to me, i think we leave blank or null by default?

File: migrations/20210914160014_group_extensions.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    })
+
+    await knex('groups').where({ type: null }).update({ type: 'group' })
```
Comment (tibetsprague) [inline migrations/20210914160014_group_extensions.js:25]: then we wouldnt need this

File: test/unit/models/GroupExtension.test.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    user = await new User({name: 'username', email: 'john@foo.com', active: true}).save()
+    await Group.create(user.id, data)
+    savedGroup = await Group.find('comm1')
+    extension = await new Extension({type: 'test'}).save()
+    groupExtension = await new GroupExtension({group_id: savedGroup.id, extension_id: extension.id, active: true, data: JSON.stringify({good: 'luck'})}).save()
+  });
+
+  after(async function() {
+    await groupExtension.destroy()
```
Comment (tibetsprague) [inline test/unit/models/GroupExtension.test.js:28]: still seems suspicious to me

File: test/unit/models/GroupExtension.test.js [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+  it('an extension can be created', function () {
+    expect(extension.id).to.exist
+  })
+
+  it('a group extension relationship is also created', function () {
+    expect(groupExtension.id).to.exist
+  })
+
+  // it('creates with default banner and avatar', async function() {
```
Comment (tibetsprague) [inline test/unit/models/GroupExtension.test.js:40]: remove

File: api/graphql/index.test.js [hunk 9]
```diff
@@ -44,6 +50,11 @@ describe('graphql request handler', () => {
     ]))
   })
 
+  after(async function() {
+    await groupExtension.destroy()
```
[the remark this answers is not in the document]
Comment (thomasgwatson) [reply api/graphql/index.test.js:54]: Indeed, this was very frustrating. The generic destruction process seemed to only work for all the other tables but groupExtension

[the remark this answers is not in the document]
Comment (tibetsprague) [reply api/graphql/index.test.js:54]: there must be something else going on here... maybe we pair on it?

File: package.json [hunk 10]
```diff
@@ -54,7 +54,8 @@
     "nock": "^7.0.2",
     "nodemon": "^1.12.1",
     "nyc": "^10.0",
-    "socket.io-client": "^2.3.1"
+    "socket.io-client": "^2.3.1",
+    "standard": "16.0.4"
```
Comment (tibetsprague) [inline package.json:58]: is this doing anything? does linting work now?

**Review by tibetsprague — APPROVED** [review]
question mark whether to merge this into dev or not... i would say lets leave on a feature branch for now until its more thoroughly tested with a completed feature. thoughts?

**Comment by thomasgwatson** [comment]
Update note: This branch has been merged into the `farm` branch and will separately be reviewed for merge into dev once that work has progressed

**Pull request closed without merging**